### PR TITLE
Divide and conquer traverse

### PR DESCRIPTION
**Pull request by hdgarrood**
Like #48, but rebased, and I fixed a minor error, and added some more tests that would have caught it. /cc @S11001001

### CODE REVIEW

File: src/Data/Traversable.js [hunk 1]
```diff
@@ -3,58 +3,32 @@
...
-            var result = go(pure(emptyList), array.length, array);
-            while (result instanceof Cont) {
-              result = result.fn();
+            function go(bot, top) {
+              switch (top - bot) {
+              case 0: return pure([]);
+              case 1: return map(array1)(f(array[bot]));
+              case 2: return apply(map(array2)(f(array[bot])))(f(array[bot + 1]));
+              default:
+                var pivot = Math.floor((top + bot) / 2);
```
Comment (hdgarrood) [inline src/Data/Traversable.js]: By the way, please feel free to take these commits and put them back into your PR if you want, and then I could close this.

Comment (hdgarrood) [inline src/Data/Traversable.js]: Or at least the ones where I added more tests.

Comment (S11001001) [inline src/Data/Traversable.js]: The idea is simply to choose pivots that will yield two even-length partitions where possible.

I suppose the best approach would be to add a `case 3`. IIUC this is the only nonterminating case under #48; 4 will be divided into 2 and 2.

**Comment by hdgarrood** [comment]
I've made the changes to pivot selection discussed earlier in this PR, and double checked that this is not a breaking change (note that we only changed `devDependencies` in bower.json). Also, now that `integers` exports an int `pow` function, I've updated the tests to use that. I think this is good to merge now.

**Comment by garyb** [comment]
Did we benchmark it at all in the end?

**Comment by hdgarrood** [comment]
We did, I added a comment in #48 with results. Performance wise it seems to be pretty much the same as the previous implementation, at least on V8.

**Comment by S11001001** [comment]
@hdgarrood I'm glad it worked; it didn't in my own tests.

The asymptotics have several dimensions, each induced by particular applicatives. Some examples:

1. A `Writer` with O(n) append will multiply out to O(n²) with the original implementation, but be O(n log﻿₂n) with divide-and-conquer.
2. A `Writer` whose monoid is fast when folding one way but slow when folding the other way will be somewhere in between.  For example, a cons list monoid with the elements more-or-less distributed evenly will be O(n) with a right-folding traversal, O(n²) with a left-folding traversal, and O(n log₂n) with divide-and-conquer. A monoid that's slow when right-folding, but fast when left-folding, also ends up with divide-and-conquer as a middle ground.
3. For `Reader` there's no difference when traversing, and no performance difference when invoking the resulting function. But *invoking* the resulting function will consume O(n) stack for the original implementation (even if "stack-safe", hence the error added by the test), but O(log₂n) stack for divide-and-conquer.  Note that to observe this, you have to invoke the function; the performance of calling `traverse` won't include this information.

**Comment by garyb** [comment]
Resolved by #78 (!)

**Pull request closed without merging**